// @aws-cdk/aws-ec2/test/test.machine-image.js
"use strict";
const core_1 = require("@aws-cdk/core");
const ec2 = require("../lib");
const lib_1 = require("../lib");
module.exports = {
    'can make and use a Windows image'(test) {
        // GIVEN
        const stack = new core_1.Stack(undefined, undefined, {
            env: { region: 'testregion' }
        });
        // WHEN
        const image = new ec2.GenericWindowsImage({
            testregion: 'ami-1234'
        });
        // THEN
        const details = image.getImage(stack);
        test.equals(details.imageId, 'ami-1234');
        test.equals(details.osType, ec2.OperatingSystemType.WINDOWS);
        test.done();
    },
    'WindowsImage retains userdata'(test) {
        // GIVEN
        const stack = new core_1.Stack(undefined, undefined, {
            env: { region: 'testregion' }
        });
        // WHEN
        const ud = ec2.UserData.forWindows();
        const image = new ec2.GenericWindowsImage({
            testregion: 'ami-1234',
        }, {
            userData: ud
        });
        // THEN
        const details = image.getImage(stack);
        test.equals(details.userData, ud);
        test.done();
    },
    'LookupMachineImage default search'(test) {
        // GIVEN
        const app = new core_1.App();
        const stack = new core_1.Stack(app, 'Stack', {
            env: { account: '1234', region: 'testregion' }
        });
        // WHEN
        new lib_1.LookupMachineImage({ name: 'bla*', owners: ['amazon'] }).getImage(stack);
        // THEN
        const missing = app.synth().manifest.missing || [];
        test.deepEqual(missing, [
            {
                key: 'ami:account=1234:filters.image-type.0=machine:filters.name.0=bla*:filters.state.0=available:owners.0=amazon:region=testregion',
                props: {
                    account: '1234',
                    region: 'testregion',
                    owners: ['amazon'],
                    filters: {
                        'name': ['bla*'],
                        'state': ['available'],
                        'image-type': ['machine']
                    }
                },
                provider: 'ami'
            }
        ]);
        test.done();
    }
};
