// @aws-cdk/core/lib/private/asset-parameters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cfn_parameter_1 = require("../cfn-parameter");
const construct_1 = require("../construct");
class FileAssetParameters extends construct_1.Construct {
    constructor(scope, id) {
        super(scope, id);
        // add parameters for s3 bucket and s3 key. those will be set by
        // the toolkit or by CI/CD when the stack is deployed and will include
        // the name of the bucket and the S3 key where the code lives.
        this.bucketNameParameter = new cfn_parameter_1.CfnParameter(this, 'S3Bucket', {
            type: 'String',
            description: `S3 bucket for asset "${id}"`,
        });
        this.objectKeyParameter = new cfn_parameter_1.CfnParameter(this, 'S3VersionKey', {
            type: 'String',
            description: `S3 key for asset version "${id}"`
        });
        this.artifactHashParameter = new cfn_parameter_1.CfnParameter(this, 'ArtifactHash', {
            description: `Artifact hash for asset "${id}"`,
            type: 'String',
        });
    }
}
exports.FileAssetParameters = FileAssetParameters;
class DockerImageAssetParameters extends construct_1.Construct {
    constructor(scope, id) {
        super(scope, id);
        this.imageNameParameter = new cfn_parameter_1.CfnParameter(this, 'ImageName', {
            type: 'String',
            description: `ECR repository name and tag for asset "${id}"`,
        });
    }
}
exports.DockerImageAssetParameters = DockerImageAssetParameters;
