// @aws-cdk/core/test/private/test.tree-metadata.js
"use strict";
const fs = require("fs");
const path = require("path");
const index_1 = require("../../lib/index");
function readJson(outdir, file) {
    return JSON.parse(fs.readFileSync(path.join(outdir, file), 'utf-8'));
}
module.exports = {
    'tree metadata is generated as expected'(test) {
        const app = new index_1.App();
        const stack = new index_1.Stack(app, 'mystack');
        new index_1.Construct(stack, 'myconstruct');
        const assembly = app.synth();
        const treeArtifact = assembly.tree();
        test.ok(treeArtifact);
        test.deepEqual(readJson(assembly.directory, treeArtifact.file), {
            version: 'tree-0.1',
            tree: {
                id: 'App',
                path: '',
                children: {
                    Tree: {
                        id: 'Tree',
                        path: 'Tree'
                    },
                    mystack: {
                        id: 'mystack',
                        path: 'mystack',
                        children: {
                            myconstruct: {
                                id: 'myconstruct',
                                path: 'mystack/myconstruct'
                            }
                        }
                    }
                }
            }
        });
        test.done();
    },
    'tree metadata for a Cfn resource'(test) {
        class MyCfnResource extends index_1.CfnResource {
            constructor(scope, id) {
                super(scope, id, {
                    type: 'CDK::UnitTest::MyCfnResource'
                });
            }
            inspect(inspector) {
                inspector.addAttribute('aws:cdk:cloudformation:type', 'CDK::UnitTest::MyCfnResource');
                inspector.addAttribute('aws:cdk:cloudformation:props', this.cfnProperties);
            }
            get cfnProperties() {
                return {
                    mystringpropkey: 'mystringpropval',
                    mylistpropkey: ['listitem1'],
                    mystructpropkey: {
                        myboolpropkey: true,
                        mynumpropkey: 50
                    }
                };
            }
        }
        const app = new index_1.App();
        const stack = new index_1.Stack(app, 'mystack');
        new MyCfnResource(stack, 'mycfnresource');
        const assembly = app.synth();
        const treeArtifact = assembly.tree();
        test.ok(treeArtifact);
        test.deepEqual(readJson(assembly.directory, treeArtifact.file), {
            version: 'tree-0.1',
            tree: {
                id: 'App',
                path: '',
                children: {
                    Tree: {
                        id: 'Tree',
                        path: 'Tree'
                    },
                    mystack: {
                        id: 'mystack',
                        path: 'mystack',
                        children: {
                            mycfnresource: {
                                id: 'mycfnresource',
                                path: 'mystack/mycfnresource',
                                attributes: {
                                    'aws:cdk:cloudformation:type': 'CDK::UnitTest::MyCfnResource',
                                    'aws:cdk:cloudformation:props': {
                                        mystringpropkey: 'mystringpropval',
                                        mylistpropkey: ['listitem1'],
                                        mystructpropkey: {
                                            myboolpropkey: true,
                                            mynumpropkey: 50
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
        });
        test.done();
    }
};
