// @aws-cdk/assert/lib/assertions/count-resources.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const assertion_1 = require("../assertion");
/**
 * An assertion to check whether a resource of a given type and with the given properties exists, disregarding properties
 */
function countResources(resourceType, count = 1) {
    return new CountResourcesAssertion(resourceType, count);
}
exports.countResources = countResources;
class CountResourcesAssertion extends assertion_1.Assertion {
    constructor(resourceType, count) {
        super();
        this.resourceType = resourceType;
        this.count = count;
        this.inspected = 0;
    }
    assertUsing(inspector) {
        let counted = 0;
        for (const logicalId of Object.keys(inspector.value.Resources || {})) {
            const resource = inspector.value.Resources[logicalId];
            if (resource.Type === this.resourceType) {
                counted++;
                this.inspected += 1;
            }
        }
        return counted === this.count;
    }
    get description() {
        return `stack only has ${this.inspected} resource of type ${this.resourceType} but we expected to find ${this.count}`;
    }
}
