// @aws-cdk/assert/test/test.have-resource.js
"use strict";
const cxapi = require("@aws-cdk/cx-api");
const fs_1 = require("fs");
const path_1 = require("path");
const index_1 = require("../lib/index");
function mkStack(template) {
    const assembly = new cxapi.CloudAssemblyBuilder();
    assembly.addArtifact('test', {
        type: cxapi.ArtifactType.AWS_CLOUDFORMATION_STACK,
        environment: cxapi.EnvironmentUtils.format('123456789', 'us-west-2'),
        properties: {
            templateFile: 'template.json'
        }
    });
    fs_1.writeFileSync(path_1.join(assembly.outdir, 'template.json'), JSON.stringify(template));
    return assembly.buildAssembly().getStackByName('test');
}
module.exports = {
    'support resource with no properties'(test) {
        const synthStack = mkStack({
            Resources: {
                SomeResource: {
                    Type: 'Some::Resource'
                }
            }
        });
        index_1.expect(synthStack).to(index_1.haveResource('Some::Resource'));
        test.done();
    },
    'haveResource tells you about mismatched fields'(test) {
        const synthStack = mkStack({
            Resources: {
                SomeResource: {
                    Type: 'Some::Resource',
                    Properties: {
                        PropA: 'somevalue'
                    }
                }
            }
        });
        test.throws(() => {
            index_1.expect(synthStack).to(index_1.haveResource('Some::Resource', {
                PropA: 'othervalue'
            }));
        }, /PropA/);
        test.done();
    },
    'haveResource value matching is strict by default'(test) {
        const synthStack = mkStack({
            Resources: {
                SomeResource: {
                    Type: 'Some::Resource',
                    Properties: {
                        PropA: {
                            foo: 'somevalue',
                            bar: 'This is unexpected, so the value of PropA doesn\'t strictly match - it shouldn\'t pass'
                        },
                        PropB: 'This property is unexpected, but it\'s allowed'
                    }
                }
            }
        });
        test.throws(() => {
            index_1.expect(synthStack).to(index_1.haveResource('Some::Resource', {
                PropA: {
                    foo: 'somevalue'
                }
            }));
        }, /PropA/);
        test.done();
    },
    'haveResource allows to opt in value extension'(test) {
        const synthStack = mkStack({
            Resources: {
                SomeResource: {
                    Type: 'Some::Resource',
                    Properties: {
                        PropA: {
                            foo: 'somevalue',
                            bar: 'Additional value is permitted, as we opted in'
                        },
                        PropB: 'Additional properties is always okay!'
                    }
                }
            }
        });
        index_1.expect(synthStack).to(index_1.haveResource('Some::Resource', {
            PropA: {
                foo: 'somevalue'
            }
        }, undefined, true));
        test.done();
    },
};
