// aws-cdk/lib/api/deployment-target.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const logging_1 = require("../logging");
const serialize_1 = require("../serialize");
const credentials_1 = require("./aws-auth/credentials");
const deploy_stack_1 = require("./deploy-stack");
const toolkit_info_1 = require("./toolkit-info");
exports.DEFAULT_TOOLKIT_STACK_NAME = 'CDKToolkit';
/**
 * Default provisioner (applies to CloudFormation).
 */
class CloudFormationDeploymentTarget {
    constructor(props) {
        this.aws = props.aws;
    }
    async readCurrentTemplate(stack) {
        logging_1.debug(`Reading existing template for stack ${stack.displayName}.`);
        const cfn = await this.aws.cloudFormation(stack.environment.account, stack.environment.region, credentials_1.Mode.ForReading);
        try {
            const response = await cfn.getTemplate({ StackName: stack.stackName }).promise();
            return (response.TemplateBody && serialize_1.deserializeStructure(response.TemplateBody)) || {};
        }
        catch (e) {
            if (e.code === 'ValidationError' && e.message === `Stack with id ${stack.stackName} does not exist`) {
                return {};
            }
            else {
                throw e;
            }
        }
    }
    async deployStack(options) {
        const toolkitInfo = await toolkit_info_1.loadToolkitInfo(options.stack.environment, this.aws, options.toolkitStackName || exports.DEFAULT_TOOLKIT_STACK_NAME);
        return deploy_stack_1.deployStack({
            stack: options.stack,
            deployName: options.deployName,
            roleArn: options.roleArn,
            notificationArns: options.notificationArns,
            quiet: options.quiet,
            sdk: this.aws,
            ci: options.ci,
            reuseAssets: options.reuseAssets,
            toolkitInfo,
            tags: options.tags
        });
    }
}
exports.CloudFormationDeploymentTarget = CloudFormationDeploymentTarget;
